File: ruby/lib/hanoi.rb
module Hanoi
  Disk = Struct.new(:size) do
    def to_s
      '**' * size
    end
  end

  Base = Struct.new(:id, :disks) do
    def self.full(id, height)
      new(id, 1.upto(height).map { |size| Disk.new(size) })
    end

    def self.empty(id)
      new(id, [])
    end

    def initialize(*)
      super
      check_disks
    end

    def empty?
      disks.empty?
    end

    def top
      raise "base #{id} is empty" if empty?
      disks.first
    end

    def pop
      self.class.new(id, disks.drop(1))
    end

    def push(disk)
      self.class.new(id, [disk] + disks)
    end

    def to_s(width = disks.map(&:to_s).map(&:length).max || 0, height = disks.length)
      (
        [' ' * width] * (height - disks.length) +
        disks.map { |disk| disk.to_s.center(width) } +
        [
          '-' * width,
          id.inspect.ljust(width)
        ]
      ).join("\n")
    end

    private

    def check_disks
      disks.each_cons(2) do |a, b|
        unless a.size < b.size
          raise "can’t put #{a} on top of #{b}"
        end
      end
    end
  end

  Tower = Struct.new(:bases, :moves) do
    PEG_COUNT = 3

    def self.create(height)
      id, *ids = (:a..).take(PEG_COUNT)
      bases = [Base.full(id, height)] + ids.map { |id| Base.empty(id) }
      new(bases, [])
    end

    def move(from, to)
      disk = bases.detect { |base| base.id == from }.top
      self.class.new(
        bases.map { |base|
          base = base.pop if base.id == from
          base = base.push(disk) if base.id == to
          base
        },
        moves + [{ from: from, to: to }]
      )
    end

    def inspect
      width = bases.flat_map(&:disks).map(&:to_s).map(&:length).max || 0
      height = bases.map(&:disks).map(&:length).sum

      "\n" +
        bases.
          map { |base| base.to_s(width, height) }.
          map { |string| string.split("\n") }.
          transpose.
          map { |strings| strings.join(' ') }.
          join("\n")
    end
  end
end

File: javascript/lib/hanoi.js
const inspect = Symbol.for('nodejs.util.inspect.custom');

export class Disk {
  constructor(size) {
    this.size = size;
  }

  toString() {
    return '**'.repeat(this.size);
  }
}

export class Base {
  constructor(id, disks) {
    this.id = id;
    this.disks = disks;
    this.checkDisks();
  }

  checkDisks() {
    for (let i = 0; i < this.disks.length - 1; i++) {
      const [a, b] = this.disks.slice(i, i + 2);
      if (a.size >= b.size) {
        throw `can’t put ${a} on top of ${b}`
      }
    }
  }

  isEmpty() {
    return this.disks.length === 0;
  }

  top() {
    if (this.isEmpty()) {
      throw `base ${this.id} is empty`;
    }

    return this.disks[0];
  }

  pop() {
    return new Base(this.id, this.disks.slice(1));
  }

  push(disk) {
    return new Base(this.id, [disk].concat(this.disks));
  }

  toString(
    width = Math.max(...this.disks.map(disk => disk.toString().length), 0),
    height = this.disks.length
  ) {
    return [
      ...new Array(height - this.disks.length).fill(' '.repeat(width)),
      ...this.disks
        .map(disk => disk.toString())
        .map(string => string.padStart((string.length + width) / 2).padEnd(width)),
      '-'.repeat(width),
      this.id.toString().padEnd(width)
    ].join('\n');
  }
}

Base.full = (id, height) => {
  const disks = [];
  for (let size = 1; size <= height; size++) {
    disks.push(new Disk(size));
  }

  return new Base(id, disks);
};

Base.empty = id => {
  return new Base(id, []);
};

export class Tower {
  constructor(bases, moves) {
    this.bases = bases;
    this.moves = moves;
  }

  move(from, to) {
    const disk = this.bases.find(base => base.id === from).top();
    const bases = this.bases.map(base => {
      let result = base;
      if (result.id === from) {
        result = result.pop();
      }
      if (result.id === to) {
        result = result.push(disk);
      }
      return result;
    });
    const moves = this.moves.concat({ from, to });

    return new Tower(bases, moves);
  }

  toString() {
    const width =
      Math.max(
        ...[].concat(...this.bases.map(base => base.disks))
          .map(disk => disk.toString().length),
        0
      );
    const height =
      this.bases
        .map(base => base.disks.length)
        .reduce((a, b) => a + b, 0);

    const strings =
      this.bases
        .map(base => base.toString(width, height))
        .map(string => string.split('\n'));

    return strings[0].map((_, index) => strings.map(row => row[index]))
      .map(strings => strings.join(' '))
      .join('\n');
  }

  [inspect]() {
    return this.toString();
  }
}

const PEG_COUNT = 3;
const ASCII_A = 97;

Tower.create = height => {
  let letters = [];
  for (let i = 0; i < PEG_COUNT; i++) {
    letters.push(String.fromCharCode(ASCII_A + i));
  }

  const [id, ...ids] = letters;
  const bases = [Base.full(id, height)].concat(ids.map(id => Base.empty(id)));

  return new Tower(bases, []);
};

const Hanoi = { Disk, Base, Tower };
export default Hanoi;
